### PR TITLE
[Notifier] Add SMSFactor bridge

### PR DESCRIPTION
**Pull request by Gwemox**
| Q             | A
| ------------- | ---
| Branch?       | 6.2
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/16901
| Recipe PR | https://github.com/symfony/recipes/pull/1102

Provides [SMSFactor](https://www.smsfactor.com/) integration for Symfony Notifier.

### CODE REVIEW

File: src/Symfony/Component/Notifier/Bridge/SmsFactor/README.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
+SMSFactor Notifier
+-------------------
+
+Provides [SMSFactor](https://www.smsfactor.com/) integration for Symfony Notifier.
+
+DSN example
+-----------
+
+```
+SMS_FACTOR_DSN=sms-factor://TOKEN@default?sender=SENDER&pushType=PUSH_TYPE
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/SmsFactor/README.md]:
```suggestion
SMS_FACTOR_DSN=sms-factor://TOKEN@default?sender=SENDER&push_type=PUSH_TYPE
```

snake case please

File: src/Symfony/Component/Notifier/Bridge/SmsFactor/README.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+ - `TOKEN` is your SMSFactor api token
+ - `SENDER` is the sender name
+ - `PUSH_TYPE` is the sms type (`alert` or `marketing`)
+
+example:
+
+```
+SMS_FACTOR_DSN=sms-factor://eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyfQ.SflKxwRJSMeKKF2QT4fwpMeJf36POk6yJV_adQssw5c@default?sender=MY_COMPANY&pushType=alert
+```
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/SmsFactor/README.md]: Can be removed

File: src/Symfony/Component/Notifier/Bridge/SmsFactor/SmsFactorPushType.php [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+namespace Symfony\Component\Notifier\Bridge\SmsFactor;
+
+/**
+ * @author Thibault Buathier <thibault.buathier@gmail.com>
+ */
+enum SmsFactorPushType: string
+{
+    case Alert = 'alert';
+
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/SmsFactor/SmsFactorPushType.php]:
```suggestion
```

File: src/Symfony/Component/Notifier/Bridge/SmsFactor/SmsFactorTransport.php [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        if (null !== $this->sender) {
+            $query['sender'] = $this->sender;
+        }
+
+        if (null !== $this->pushType) {
+            $query['pushtype'] = $this->pushType->value;
+        }
+
+        $response = $this->client->request('GET', 'https://'.$this->getEndpoint().'/send', [
+            'query' => $query,
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/SmsFactor/SmsFactorTransport.php:86]: Query is correct?

File: src/Symfony/Component/Notifier/Bridge/SmsFactor/SmsFactorTransportFactory.php [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+        return (new SmsFactorTransport($tokenApi, $sender, $pushType, $this->client, $this->dispatcher))->setHost($host)->setPort($port);
+    }
+
+    protected function getSupportedSchemes(): array
+    {
+        return ['sms-factor'];
+    }
+
+    protected function getPushType(Dsn $dsn): ?SmsFactorPushType
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/SmsFactor/SmsFactorTransportFactory.php]:
```suggestion
    private function getPushType(Dsn $dsn): ?SmsFactorPushType
```
?

File: src/Symfony/Component/Notifier/Bridge/SmsFactor/composer.json [hunk 6]
```diff
@@ -0,0 +1,31 @@
+{
+    "name": "symfony/smsfactor-notifier",
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/SmsFactor/composer.json]:
```suggestion
    "name": "symfony/smsfactor-notifier",
```
@fabpot are you fine with the package name?

File: src/Symfony/Component/Notifier/Bridge/SmsFactor/composer.json [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    "type": "symfony-notifier-bridge",
+    "description": "Symfony SMSFactor Notifier Bridge",
+    "keywords": [ "sms", "smsfactor", "notifier"],
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/SmsFactor/composer.json]:
```suggestion
    "keywords": ["sms", "smsfactor", "notifier"],
```

File: src/Symfony/Component/Notifier/Bridge/SmsFactor/composer.json [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+        },
+        {
+            "name": "Symfony Community",
+            "homepage": "https://symfony.com/contributors"
+        }
+    ],
+    "require": {
+        "php": ">=8.1",
+        "symfony/http-client": "^5.4|^6.0",
+        "symfony/notifier": "^5.4|^6.0",
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/SmsFactor/composer.json:21]: I think this does not work with notifier 5.4, must be 6.2

**Review by OskarStark — COMMENTED** [review]
Thanks for the new bridge 😃

File: src/Symfony/Component/Notifier/Bridge/SmsFactor/SmsFactorTransport.php (hunk 4 above)
Comment (Gwemox) [reply src/Symfony/Component/Notifier/Bridge/SmsFactor/SmsFactorTransport.php:86]: SMSFactor requests the information in query parameters. Cf https://dev.smsfactor.com/en/api/sms/send/send-single

File: src/Symfony/Component/Notifier/Bridge/SmsFactor/composer.json (hunk 8 above)
Comment (Gwemox) [reply src/Symfony/Component/Notifier/Bridge/SmsFactor/composer.json:21]: How can I test this?

File: src/Symfony/Component/Notifier/Bridge/SmsFactor/composer.json (hunk 6 above)
Comment (fabpot) [reply src/Symfony/Component/Notifier/Bridge/SmsFactor/composer.json]: Must be `sms-factor-notifier` as the class name is `SmsFactor` and not `Smsfactor`

**Review by fabpot — CHANGES_REQUESTED** [review]

**Review by OskarStark — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @Gwemox.

**Pull request merged**